extensions/reactive-db2-client/deployment/src/main/java/io/quarkus/reactive/db2/client/deployment/ReactiveDB2ClientProcessor.java

```diff
@@ -30,9 +31,24 @@
 
 class ReactiveDB2ClientProcessor {
 
+    /**
+     * The producer needs to be produced in a separate method to avoid a circular dependency (the Vert.x instance creation
+     * consumes the AdditionalBeanBuildItems).
+     */
     @BuildStep
-    AdditionalBeanBuildItem registerBean() {
-        return AdditionalBeanBuildItem.unremovableOf(DB2PoolProducer.class);
+    void poolProducer(
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans,
+            DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
+            return;
+        }
+
+        additionalBeans.produce(new AdditionalBeanBuildItem.Builder()
+                .addBeanClass(DB2PoolProducer.class)
+                .setUnremovable()
+                .setDefaultScope(DotNames.APPLICATION_SCOPED)
```

Comment:
Is this the default when no scope is used?

Reply:
Yeah, well, I don't know but if the default changes one day, at least it will be OK.

Reply:
But does this change, change what it **is** today?

Reply:
Ah OK, I misunderstood your question. Yes, exactly.

I always prefer to remove the scope annotation from the additional beans in case at some point we index the extensions.

Reply:
> I always prefer to remove the scope annotation from the additional beans in case at some point we index the extensions.

That would not help - a class that contains a producer is automatically picked up. If a bean should be `@ApplicationScoped` then the class should IMO contain the annotation. The primary purpose of `AdditionalBeanBuildItem` is to register a class for discovery (i.e. take it as it is). AFAIR the ``AdditionalBeanBuildItem``s default scope was intended for the cases where we need to change the scope for a class that we don't control.

NOTE: In this particular case we should consider making the bean `@Dependent` as the produced `io.vertx.mutiny.db2client.DB2Pool` is singleton, i.e. the producer method is only called once, i.e. this should work fine...

```java
@Dependent
public class DB2PoolProducer {

    /**
     * @return the <em>mutiny</em> DB2 Pool instance. The instance is created lazily.
     */
    @Singleton
    @Produces
    public io.vertx.mutiny.db2client.DB2Pool mutinyDB2Pool(DB2Pool db2Pool) {
        return io.vertx.mutiny.db2client.DB2Pool.newInstance(db2Pool);
    }
}
```

Reply:
Also ``AdditionalBeanBuildItem``s default scope requires unnecessary transformation and a special "stereotype hack". So it's always better to avoid it (if possible)...

Reply:
@gsmet @geoand ^

Reply:
Good to know @mkouba, thanks!